simulator/src/testbench.js

```diff
@@ -16,20 +17,143 @@ const CONTEXT = {
     CONTEXT_ASSIGNMENTS: 1,
 };
 
-/**
- * @typedef {number} Mode
- */
-const MODE = {
-    MODE_RUNALL: 0,
-    MODE_MANUAL: 1,
-};
 
 // Do we have any other function to do this?
 // Utility function. Converts decimal number to binary string
 function dec2bin(dec) {
     return (dec >>> 0).toString(2);
 }
 
+/**
+ * Class to store all data related to the testbench and functions to use it
+ * @param {Object} data - Javascript object of the test data
+ * @param {number=} currentGroup - Current group index in the test
+ * @param {number=} currentCase - Current case index in the group
+ */
+export class TestbenchData {
+    constructor(data, currentGroup = 0, currentCase = 0) {
+        this.currentCase = currentCase;
+        this.currentGroup = currentGroup;
+        this.testData = data;
+    }
+
+    /**
+     * Checks whether given case-group pair exists in the test
+     */
+    isCaseValid() {
+        if(this.currentGroup >= this.data.groups.length || this.currentGroup < 0) return false;
+        const caseCount = this.testData.groups[this.currentGroup].inputs[0].values.length;
+        if(this.currentCase >= caseCount || this.currentCase < 0) return false;
+
+        return true;
+    }
+
+    /**
+     * Validate and set case and group in the test
+     * @param {number} groupIndex - Group index to set
+     * @param {number} caseIndex -  Case index to set
+     */
+    setCase(groupIndex, caseIndex) {
+        const newCase = new TestbenchData(this.testData, groupIndex, caseIndex);
```

Comment:
Creating a new object here seems odd?

Reply:
That object is used to do temporary operations on to check if the case and group are valid, if they are valid then those changes are made to the real object